### ISSUE TITLE
Invalid character found in the request target. The valid characters are defined in RFC 7230 and RFC 3986

### ISSUE THREAD

**Issue by andersonkxiass** [bohnman/squiggly-java#42]
I'm getting the following error, when I use `field[sub_field]` pattern:

```
	java.lang.IllegalArgumentException: Invalid character found in the request target. The valid characters are defined in RFC 7230 and RFC 3986
	at org.apache.coyote.http11.Http11InputBuffer.parseRequestLine(Http11InputBuffer.java:479) ~[tomcat-embed-core-8.5.31.jar:8.5.31]
	at org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:687) ~[tomcat-embed-core-8.5.31.jar:8.5.31]
	at org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:66) [tomcat-embed-core-8.5.31.jar:8.5.31]
	at org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:790) [tomcat-embed-core-8.5.31.jar:8.5.31]
	at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1468) [tomcat-embed-core-8.5.31.jar:8.5.31]
	at org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:49) [tomcat-embed-core-8.5.31.jar:8.5.31]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [na:1.8.0_172]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [na:1.8.0_172]
	at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61) [tomcat-embed-core-8.5.31.jar:8.5.31]
	at java.lang.Thread.run(Thread.java:748) [na:1.8.0_172]
```

**Comment by bohnman** [comment]
Can you provide the exact filter string you are using?

**Comment by bohnman** [comment]
Well shoot.  It looks like Tomcat increased their security yet again in version 8.5.31  and no longer allows raw square brackets in the query string.

Here are the options you have:

1) encode the square brackets: ``field%5Bsub_field%5B``
2) use dot notation: field.sub_field1,field.sub_field2
3) use an older version of Tomcat 8.5.30 and below
4) enable raw curly braces.

```java
System.setProperty("tomcat.util.http.parser.HttpParser.requestTargetAllow", "{}");
```

Then you can do
``field{sub_field}``

A couple notes on item 4:
- Tomcat does not allow you to do this for square brackets.
- enabling these raw fields leaves you server vulnerable to request smuggling.  Please see this [CVE](https://access.redhat.com/security/cve/cve-2016-6816) for more information and determine if this applies to you.

**Comment by rkcreddyt** [comment]
We found the same issue . After analyzing the issue , found the solution . Below are the details for issue and solution.

Issue: Tomcat (7.0.88) is throwing below exception which leads to 400 – Bad Request. java.lang.IllegalArgumentException: Invalid character found in the request target. The valid characters are defined in RFC 7230 and RFC 3986. This issue is occurring most of the tomcat versions from 7.0.88 onwards.

Solution: (Suggested by Apache team):

Tomcat increased their security and no longer allows raw square brackets in the query string. In the request we have [,] (Square brackets) so hotel request is not processing by server.
As a solution below are the steps needs to be incorporate to make it work as expected. 1. We need to add relaxedQueryChars attribute under tag under server.xml (%TOMCAT_HOME%/conf)

          <Connector port="80" 
           protocol="HTTP/1.1"
           maxThreads="150"
           connectionTimeout="20000"
           redirectPort="443"
           compression="on"
           compressionMinSize="2048"
           noCompressionUserAgents="gozilla, traviata"
           compressableMimeType="text/html,text/xml"
                                     relaxedQueryChars="[,]"
             />
If application needs more special characters which is not supported by tomcat by default then we need add those special characters in relaxedQueryChars attribute with comma separated as above.

**Comment by bohnman** [comment]
Thanks for the info!

Just to let you know, Squiggly 2 will have an alternative syntax, which will be more url-friendly.

**Comment by ahoffskov** [comment]
> relaxedQueryChars="[,]"

Solved my problem with querying an ERDDAP server. Thank you very much :-)

**Comment by xak2000** [comment]
@bohnman 

> 1. encode the square brackets: `field%5Bsub_field%5B`

Small fix: `%5B` and `%5D`.

And, just a question: if we need to encode the symbols anyway, why not use `{` (`%7B`) and `}` (`%7D`) instead of `[` (`%5B`) and `]` (`%5D`)? What type of braces should be preferred if we must encode them anyway? I understand that `squiggly` support both, but still interested if there any reason to prefer one to the other.

> Just to let you know, Squiggly 2 will have an alternative syntax, which will be more url-friendly.

Where can I find any info about new syntax?

**Comment by bohnman** [comment]
@xak2000 Good catch!  I have updated my comment.  There's no reason to prefer square brackets over curly braces.  Historically, when Squiggly was first released, Tomcat allowed curly braces to be unescaped.  Then, they forced escaping curly braces so I switched to square brackets, which they subsequently forbid.

The new syntax will use parentheses like so:

```
address(city,state,zip)
```

The best way to find info on the new syntax is to look at the SquigglyGrammar.g4 file in the 2.x branch.  Be warned that this branch doesn't even compile.  Proceed at your own risk :-)
